### ISSUE TITLE
KafkaUnavailableError

### ISSUE THREAD

**Issue by mszadig** [ciena/afkak#47]
The afkak client and producer are used for opencord/voltha kafka access.   We are seeing every 10+ minutes the following error:Unable to load metadata from configured hosts: <twisted.python.failure.Failure afkak.common.KafkaUnavailableError: All servers ([<KafkaBrokerClient 192.168.0.202:9092 clientId=u'afkak-client' connected>]) failed to process request>

When the error occurs there is no indication in the kafka-docker logs of any problem/error.
When this occurs the voltha/kafka_proxy stop /restart is generated which results in:

[2018-10-03 17:49:11,552] INFO [GroupMetadataManager brokerId=1001] Removed 0 expired offsets in 1 milliseconds. (kafka.coordinator.group.GroupMetadataManager)

Which is an expected message.

This error occurs when using afkak 3.0.0.dev0 and 2.4.0.

**Comment by twm** [comment]
Usually "Unable to load metadata from configured hosts" means the Kafka cluster is in bad shape. It's difficult to infer more without the specific error code that came back from the broker (should be logged somewhere above the KafkaUnavailableError).

BTW master (3.0.0.dev0) is in flux right now: I don't recommend using it unless you want to do early testing of the Python 3 port.

**Comment by rthille** [comment]
If it's every 10 minutes, it could possibly be due to the Kafka Broker idle disconnect.  It looks like you've got a single broker, so when Afkak detects the disconnect, it'll try to send a metadata request to any remaining brokers first, then to the ones which were in the disconnected state.  If there's a bug in Afkak wrt handling the reconnection logic when sending packets to the disconnected broker, or if re-establishing the connection just takes too long due to the broker being under load or on slow hardware, that could result in that error.

**Comment by rthille** [comment]
If you can reproduce it reliably, running Afkak with debug logging turned out would allow me to help diagnose exactly what's going on. But admittedly, Afkak does log a s**t-ton in debug mode. :-)

**Comment by mszadig** [comment]
We are seeing this problem on both a single and 3 broker set up.    I have the clients publishing to kafka every 200ms, so I would not expect the broker to be idle.  What do I do to turn on debug more for afkak?

**Comment by twm** [comment]
Afkak uses the stdlib [logging package](https://docs.python.org/2/library/logging.html), so you can capture debug logs by adding a handler to the "afkak" logger. A primitive example:

```python
import logging
logging.basicConfig()
```

If you've suppressed Afkak's debug logs (e.g. by setting its logger level to INFO) you'll want to set it to debug:

```python
logging.getLogger('afkak').setLevel(logging.DEBUG)
```

**Comment by mszadig** [comment]
I am having difficultly getting logging.DEBUG running.  
Our configuration runs in a container that has loaded afkak in the dist-packages of usr/local/lib/python2.7.    The module that calls the afkak.producer,client and common libraries is running structlog instead of the standard library.   Do you have any suggestions regarding how to enable the debug level with structlog being used to set logging levels for the other modules?

I have also verified that KafkaUnavailable problem exists when using 2.4.0.

**Comment by twm** [comment]
I'm not really familiar with structlog, but it looks like you need to configure the logging library _in addition_ to it: https://structlog.readthedocs.io/en/latest/standard-library.html

**Comment by mszadig** [comment]
Still having trouble getting the logging to function properly but when the error occurs, the brokerclient reports:  [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
]
Does this provide any insight into what is happening?

**Comment by rthille** [comment]
One thing you could maybe do to get more insight is to run `netstat -nlp | grep <PID-of-process-with-afkak-client>` in a loop and see if the connection is dropped every ~10 minutes.  `tcpdump` may also be useful if the data volume isn't too large on the connection as we can then see which end (afkak vs Kafka) is initiating the close of the connection.

**Comment by rthille** [comment]
Another option, if you can would be to share the Dockerfile, if the code isn't proprietary. That might allow me to reproduce.

**Comment by mszadig** [comment]
I can perform the suggestion above.  But I have determined that the problem is in the afkak library.  The cord/voltha code wraps the afkak library in a proxy:  https://github.com/opencord/voltha/blob/master/voltha/northbound/kafka/kafka_proxy.py

I substituted the imports for afkak with imports for the kafka-python package https://pypi.org/project/kafka-python/   and the code ran without any errors or disconnects.   I monitored the kafka outputs using kafkacat.   All kafka send   messages were sent and consumed by kafkacat.

The kafka docker file is called by kubermetes node via https://github.com/opencord/voltha/blob/master/k8s/kafka.yml
The actually docker file is wurstemeister/1.0.0.   I have tried more recent container versions with the same results using afkak.

If you want the tcpdump that I ran last night prior to substituting the package kafka-python, I can send that to you.   It is also quite easy to switch back to the afkak library, if you want to pursue this.
i pulled the afkak library directly into my dev env. with https://github.com/ciena/afkak.git  --branch 2.4.0.

**Comment by rthille** [comment]
Sure, you can either email the tcpdump to me, or post it somewhere and drop a link here.  org-afkak-20181017@rangat.org works for an email address.

Odd that Kafka-python would work, given it's not Twisted, so it'll block the reactor.  I guess it's just producing, so it's unlikely to block the reactor for too long.

**Comment by mszadig** [comment]
I sent the dump file.   Let me know if you need more info.

**Comment by rthille** [comment]
Ok, it's "weird" :-). Everything seems to be going along fine until fragment number 271 @ time 530.026202, when Kafka closes the connection we haven't seen in the trace before that (192.168.0.204:9092->192.168.0.206:42414). All the previous packets were between client port 42416 and 9092.  Immediately after that, Afkak appears to close the connection all the traffic had been on (192.168.0.204:9092<->192.168.0.206:42416).
Then it goes into a loop where a new connection to Kafka is opened, a metadata request is sent, the response is received, and then the connection is closed.
It loops for about 10 seconds, getting the metadata over and over again, then there's a RST, which I'm thinking is just dumb luck, but oddly there's only one more short-connection to get metadata and then a new connection is opened from 47484 and immediately a Produce request is sent with no metadata request.  That connection is used for about 10 minutes and then an idle connection from 47482 is closed and it all starts over again with the looping of new connections to get metadata and that continues until we see a couple more RSTs and then we're producing data again.

So, one thing that might be going on is that Afkak is getting the broker metadata and trying to connect to the broker at that DNS name (kafka-0.kafka.voltha.svc.cluster.local:9092), but that's not the name it's been configured with, so it's getting confused?

**Comment by rthille** [comment]
@twm This is the line I'd worry most about in that change: https://github.com/ciena/afkak/compare/2.3.0...2.4.0#diff-edf761ee6861903619be6edbe92e7d69R684
That might be triggering a bug in this code: https://github.com/ciena/afkak/compare/2.3.0...2.4.0#diff-edf761ee6861903619be6edbe92e7d69R579

**Comment by rthille** [comment]
Gah, neither of those links seem to work...

**Comment by rthille** [comment]
`https://github.com/ciena/afkak/compare/2.3.0...2.4.0#diff-edf761ee6861903619be6edbe92e7d69R684`
`https://github.com/ciena/afkak/compare/2.3.0...2.4.0#diff-edf761ee6861903619be6edbe92e7d69R579`
Somehow the hash-character was removed.

**Comment by mszadig** [comment]
I will try 2.3.0 and let you know.....

**Comment by mszadig** [comment]
Robert,  This failed as well.   I pulled in the 2.3.0 version and incorporated it directly into my dev env to be certain that the correct version was running, I added to the client a self.version = "2.3.0" and log that out on the error event from the container.    For 2.3.0, The message still returns: 
error: Unable to load metadata from configured hosts: <twisted.python.failure.Failure afkak.common.KafkaUnavailableError: All servers [[('kafka-0.kafka.voltha.svc.cluster.local', 9092)]] failed to process request>

Do you want me to try something else?   I also pulled down the confluent-kafka-python and that works (I do not see the exception, although it also is not a twisted implementation).

**Comment by rthille** [comment]
I think the best bet to understand what's going wrong in Afkak is debug logging.  If you can add:
```
import logging
afkak_log = logging.getLogger('afkak')
afkak_log.setLevel(logging.DEBUG)
```
Around here: https://github.com/opencord/voltha/blob/master/voltha/northbound/kafka/kafka_proxy.py#L31
Then I *think* that should cause afkak to log at DEBUG level, despite the structlog stuff.

**Comment by mszadig** [comment]
OK.   I will get back to you soon.

**Comment by mszadig** [comment]
import logging
afkak_log = logging.getLogger('afkak')
afkak_log.setLevel(logging.DEBUG)

Has no effect.   I found this to be the case when I tried enabling logging previously.   When I set a breakpoint on the logging in your classes, the log.disabled = True and propagate is false.  If I set these directly in the classes then I can get logging to the container log file.  There must be some interaction with structlog.

Given the above I do have logging going to a file and to the log file of the docker container that is running the code.

What code base do you want me to be using.   2.3.0 or 2.4.0

**Comment by rthille** [comment]
Ah, the having to add the logging code directly accounts for the line number differences I'm seeing.

First off, is this machine massively overwhelmed?
```
20181018T200602.019 DEBUG    MainThread client._update_brokers - [line 577] - <KafkaClient clientId=afkak-client brokers=[] timeout=10.0>: _update_brokers: [('192.168.0.110', 9092)] remove: True
20181018T200609.797 DEBUG    MainThread client._get_brokerclient - [line 539] - <KafkaClient clientId=afkak-client brokers=[] timeout=10.0>: creating new KafkaBrokerClient: ('192.168.0.110', 9092)
```
Those two log messages are >7 seconds apart, but as near as I can tell there are only about 15 lines of python code between them, and nothing blocking or IO related (other than writing the log messages).

But that's not the issue, this is:
Client starts up, configured with one broker: 192.168.0.110.
```
20181018T200550.940 DEBUG    MainThread client.load_metadata_for_topics - [line 211] - <KafkaClient clientId=afkak-client brokers=[] timeout=10.0>: load_metadata_for_topics: ('voltha.events',)
20181018T200602.019 DEBUG    MainThread client._update_brokers - [line 577] - <KafkaClient clientId=afkak-client brokers=[] timeout=10.0>: _update_brokers: [('192.168.0.110', 9092)] remove: True
20181018T200609.797 DEBUG    MainThread client._get_brokerclient - [line 539] - <KafkaClient clientId=afkak-client brokers=[] timeout=10.0>: creating new KafkaBrokerClient: ('192.168.0.110', 9092)
20181018T200610.020 DEBUG    MainThread client._handleMetadataResponse - [line 229] - <KafkaClient clientId=afkak-client brokers=[('192.168.0.110', 9092)] timeout=10.0>: Broker/Topic metadata: {1001: BrokerMetadata(node_id=1001, host='kafka-0.kafka.voltha.svc.cluster.local', port=9092)}/{'voltha.events': TopicMetadata(topic='voltha.events', topic_error_code=5, partition_metadata={})}
```
And then the consumer has the client get metadata for its topic. Since the client is only getting metadata for that one topic, it doesn't remove any other brokers, it just adds the one returned: 'kafka-0.kafka.voltha.svc.cluster.local'
... Then, 10 minutes later, Kafka closes the idle connection and Afkak closes the *wrong* broker ...
```
20181018T201625.751 DEBUG    MainThread client.load_metadata_for_topics - [line 211] - <KafkaClient clientId=afkak-client brokers=[('192.168.0.110', 9092), ('kafka-0.kafka.voltha.svc.cluster.local', 9092)] timeout=10.0>: load_metadata_for_topics: ()
20181018T201625.766 DEBUG    MainThread client._update_brokers - [line 577] - <KafkaClient clientId=afkak-client brokers=[('192.168.0.110', 9092), ('kafka-0.kafka.voltha.svc.cluster.local', 9092)] timeout=10.0>: _update_brokers: [('192.168.0.110', 9092)] remove: True
```
^^That says the client has two active brokers, the 'kafka-0.kafka.voltha.svc.cluster.local' one, and the configured '192.168.0.110' one, and then it decides that it needs to update its list to *just* '192.168.0.110', and remove any others.

So, a work around would be to make sure that the client is initially configured with what the Brokers return for their names: 'kafka-0.kafka.voltha.svc.cluster.local'.

I'll have to look later and see if I can figure out why it's closing the wrong brokerClient.  Or maybe @twm can take a look, since he's still paid to maintain Afkak :-)

**Comment by mszadig** [comment]
Thanks.   I will look into the work around for now.     
To answer your question regarding the machine loading, the machine being used is a fully loaded DL360 running kubermetes.   It is lightly loaded.   There are other developers and test machines in our network and those machines are seeing the same errors that I have been reporting.   I don't think loading is a problem at this time.  There is very little going on (just heartbeat, kpis and event reporting.

The 'two' brokers are just one where the kafka-0.kafka.voltha.svc.cluster.local and 192.168.0.110 represent the same container in kubermetes  (kafka-0) to which the client is attaching.

**Comment by mszadig** [comment]
Note that for each new instance of the kubermetes node the ip will change but the dns name kafka-0.kafka.voltha.svc.cluster.local  will be the same.

**Comment by mszadig** [comment]
Where is the ip getting inserted.   When I initialize the client only the dns name is passed which sets the hosts value.   Once the messages start being sent the client reports the following brokers / hosts:
kafka_hosts: kafka.voltha.svc.cluster.local, brokers: {('192.168.0.209', 9092): <KafkaBrokerClient 192.168.0.209:9092:afkak-client, ('kafka-0.kafka.voltha.svc.cluster.local', 9092): <KafkaBrokerClient kafka-0.kafka.voltha.svc.cluster.local:9092:afkak-client}

Kafka itself  (using kafkacat)  reports:
kafkacat -b kafka.voltha.svc.cluster.local -L
Metadata for all topics (from broker -1: kafka.voltha.svc.cluster.local:9092/bootstrap):
 1 brokers:
  broker 1001 at kafka-0.kafka.voltha.svc.cluster.local:9092
 3 topics:
  topic "voltha.events" with 1 partitions:
    partition 0, leader 1001, replicas: 1001, isrs: 1001
  topic "voltha.kpis" with 1 partitions:
    partition 0, leader 1001, replicas: 1001, isrs: 1001
  topic "voltha.heartbeat" with 1 partitions:
    partition 0, leader 1001, replicas: 1001, isrs: 1001

Notice that both kafka-0.kafka.voltha.svc.cluster.local and kafka.voltha.svc.cluster.local are being reported.  They both resolve to the same ip address which in this case is 192.168.0.209 (but will change on the next restart of the kafka container).  

With kubermetes the ip will always be different for each restart.    Where do you suggest I implement the workaround.  In the current code (kafka_proxy) I can only pass in the address of the  kafka server container.   The afkak code appears to query that server to resolve the brokers and gets back 3 addresses (all of which represent a single broker) and interprets that as 3 different brokers.

Let me know if you think a workaround which addresses this is reasonable, without changing afkak code alone.

thanks

**Comment by mszadig** [comment]
Here are the results for the workaround.  
1.  I tried modifying the host name for the value sent to the client kafka.voltha.svc.cluster.local =>kafka-0.kafka.voltha.svc.cluster.local.   This continues to result in multiple broker addresses resolving to the same ip.
2.  I resolved kafka-0.kafka.voltha.svc.cluster.local to the actual ipaddr and passed that to the client init.   Same error is generated after 10 minutes.   but.... I added logging to the send message section of our code and this is reported:  kafka_hosts: 192.168.0.72, brokers: {('192.168.0.72', 9092):  There are not multiple entries for the brokers as in previous tests.

-m

**Comment by mszadig** [comment]
If I read your code correctly, the send_message routine batches the messages until the send message count is reached (default appears to be 10 messages).   This would mean the failure we are seeing there is the likelihood that 10 messages could be lost.   Is there a way to retrieve the messages that were not sent, so that on a restart these messages could be resent.    If I could grab that set then I could create a failed_message_set list and send that in on restart.    Is there a way to do this in 2.4.0?    Would grabbing producer._batch_reqs and resending them on a restart seem reasonable or would I still be in the position of losing messages? OR is there a call back that would give me the failed message set on error?

Thanks,

**Comment by twm** [comment]
I think that if you want to re-send the messages outside of Producer's automatic retries you should maintain an external data structure with the messages you have passed to `send_messages()`, removing them from there until the `Deferred` it returns succeed.s

**Comment by rthille** [comment]
Pretty sure batching is off by default.

**Comment by twm** [comment]
As for when the issue you identified will be fixed, I'm afraid I don't have a timeline for you. My priorities right now are:

* Merging coordinated consumer support (#37).
* Implementing Message/MessageSet v1 (timestamps).

**Comment by twm** [comment]
I spent some time today working on this. I'll outline the nature of the changes I'm thinking are necessary so that Robert can review:

1. `KafkaClient.clients`, which currently maps broker (host, port) to an object which maintains a connection, should be keyed by broker node ID instead of (host, port).
2. Right now `KafkaClient` eagerly populates the clients map based on the hosts it is given in the constructor.
3. The broker node IDs cannot be known until a metadata request succeeds (any broker can answer such a request).
4. There is no way (that I can find) to ask a broker you are currently connected to what its own node ID is.

Therefore:

* We must make a temporary connection to make the initial metadata request.
* If we wish to avoid loading metadata for all topics, this request must be delayed until `load_metadata_for_topics()` is first called.

We could also reduce resource use on large clusters by creating connections to brokers lazily, rather than eagerly.

**Comment by rthille** [comment]
I agree with #1.
I don't think #2 is accurate.  On `master` (if I'm reading it right), the KafkaClient constructor just stores the list of hosts (brokers) and sets a marker indicating that the hosts (may) need to be resolved.
#3. I think is the crux of the issue. We don't make full-metadata requests unless needed (connection dropped) because we've found that for clusters with lots of topics/partitions it can be **very** expensive and completely bog down the cluster.
However, we could track whether a particular KafkaBrokerClient is for the initial "hosts" connection, and/or created because of Kafka metadata which was returned.  In the case of a 'startup-only' KafkaBrokerClient connection dropping, we can ignore that until all the KafkaBrokerClients have lost their connections and we can't reconnect.  Right now, any time a connection is dropped, we naively just forget all our metadata and kick off a request for new metadata.  That's not a great approach.
#4 seems accurate.  I find it funny that they are up to v6 on the metadata response... I'm guessing the docs lie and the broker data isn't for *all* brokers, but only the brokers which carry the topic(s) being queried.

**Comment by twm** [comment]
You're right, the "eagerness" of `_KafkaBrokerClient`s (in that they attempt to maintain a connection to the broker even when it is not in use) isn't relevant here. What's directly causing this bug is that it's mixing keys from two sources.

Here `KafkaClient.clients` is updated based on the hosts:

https://github.com/ciena/afkak/blob/8cf2301e33a57ef6f26b53c7ff4af8d5918e9f13/afkak/client.py#L783-L798

And here it is updated based on the  metadata response:

https://github.com/ciena/afkak/blob/8cf2301e33a57ef6f26b53c7ff4af8d5918e9f13/afkak/client.py#L302-L310

Incidentally I think that the eagerness of `_KafkaBrokerClient` is driving some complexity here: if connections were lazily created we wouldn't need this logic to prune stale instances from the map (which won't even work unless someone does a full metadata fetch!).

**Comment by rthille** [comment]
It'd be interesting to see if there were a performance difference if the KafkaBrokerClients were to delay connecting until needed...
I think KafkaClient should just notice the disconnect, and not reset its metadata (at least not on a clean close from Kafka), and should only fall back to the configured (vs metadata acquired) BrokerClients in the case a fetch_metadata request fails.

**Comment by twm** [comment]
Yeah the current behavior on connection loss &mdash; resetting all metadata and doing a fetch of the metadata for _all_ topics &mdash; seems way too aggressive. You can lose connectivity for plenty of reasons that aren't related to brokers changing IPs.

#51 is a work-in-progress attempt to address this.

**Comment by wiml** [comment]
I am having what I think is the same problem (using afkak 3.0.0.dev20181218, twisted 18.9, and python 3.6); the symptom I see is that afkak is spinning in a loop re-resolving the introducers' IP addresses, dropping connections, and re-requesting the topic metadata. The `KafkaUnavailableError` seems to be a side effect of the rapid dropping and reconnecting.

I'm pretty sure this is triggered by one of the brokers cleanly closing an idle connection, but once it starts the connection dropping is entirely on afkak's side.

Neither the client nor the Kafka cluster are very heavily loaded at all (dev environment).

Here's a log of the start of the problem (added a handler to the afkak logger at level DEBUG, intermingled with Twisted's logging): https://gist.github.com/wiml/725b63d6725e38cac59793bff690de09

**Comment by twm** [comment]
@wiml Does it still occur with the changes on #51? The current tip (97ac499b8c5f42777241b3572a6c36c2d6b2a925) would be:

```
pip install 'https://github.com/ciena/afkak/archive/97ac499b8c5f42777241b3572a6c36c2d6b2a925.tar.gz'
```

**Comment by wiml** [comment]
It seems to be fixed in `97ac499`!

(I also checked the current tip of master, `43df374`; it still occurs there but I think that is expected, since that rev doesn't include the key_by_broker_id fixes.)

**Comment by twm** [comment]
Excellent! I'm just waiting on review there. Then I'll kick out a 3.0.0.
